man/leapctl.8.ronn
leapctl(8) -- Administrative utility for privleap
=============================================

<!--
# copyright (c) 2025 - 2025 encrypted support llc <adrelanos@whonix.org>
# see the file copying for copying conditions.
-->

## SYNOPSIS
`leapctl (--create|--destroy|--reload) [username|uid]`

## DESCRIPTION
Provides administrative controls for privleap. This is used to open and close
sockets used to allow individual user accounts to communicate with the
privleapd server. It can also tell privleapd to dynamically reload its
configuration.

You usually will not need to run this directly - if `privleap` was installed
from your distribution's repository, it should be pre-configured to create and
destroy user communication sockets on user login and logout.

## OPTIONS

* `--create` :
  Instructs privleapd to create a communication socket for the specified user.
  This will allow the user to communicate with privleapd and request
  privileged operations to be run.

* `--destroy` :
  Instructs privleapd to destroy a communication socket for the specified
  user. This will prevent the user from communicating with privleapd any
  further.

* `--reload` :
  Instructs privleapd to reload its configuration files without restarting.

## ENVIRONMENT VARIABLES
None.

## CONFIG FILES
See privleap-conf.d(5).

## EXIT CODES
0: The requested operation was successful, or did not need to be done.

1: The requested operation failed.

2: privleapd was requested to create a socket for a user, but the specified
user is not permitted to have a comm socket.

## EXAMPLES
Create a communication socket for user `user`:

`leapctl --create user`

Destroy a communication socket for the user with UID 1001:

`leapctl --destroy 1001`

## NOTES
leapctl communicates with privleapd using a UNIX socket that by default can only
be read from and written to by `root`. You will therefore usually have to run
leapctl as root for it to work.

## WWW
https://github.com/Kicksecure/privleap

## AUTHOR
This man page has been written by Aaron Rainbolt (arraybolt3@gmail.com).


debian/changelog
privleap (3:6.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sun, 06 Sep 2026 16:46:26 +0000

privleap (3:6.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 15 Aug 2026 16:36:30 +0000

privleap (3:6.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 15 Aug 2026 16:12:17 +0000

privleap (3:6.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 28 Jul 2026 13:19:43 +0000

privleap (3:6.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 21 Jul 2026 09:22:39 +0000

privleap (3:6.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 08 Jul 2026 18:12:37 +0000

privleap (3:6.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 07 Jul 2026 10:40:33 +0000

privleap (3:5.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 29 Jun 2026 15:23:51 +0000

privleap (3:5.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 28 May 2026 19:31:57 +0000

privleap (3:5.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 20 May 2026 22:16:11 +0000

privleap (3:5.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 11 May 2026 20:28:20 +0000

privleap (3:5.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 08 Apr 2026 17:49:52 +0000

privleap (3:5.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 19 Mar 2026 18:29:50 +0000

privleap (3:5.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 27 Feb 2026 10:53:43 +0000

privleap (3:5.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 28 Jan 2026 06:44:58 +0000

privleap (3:5.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 08 Jan 2026 09:43:19 +0000

privleap (3:5.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 03 Jan 2026 13:08:58 +0000

privleap (3:4.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 01 Jan 2026 17:13:55 +0000

privleap (3:4.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sun, 28 Dec 2025 06:35:59 +0000

privleap (3:4.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 03 Dec 2025 08:30:23 +0000

privleap (3:4.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 19 Nov 2025 07:06:51 +0000

privleap (3:4.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 18 Nov 2025 07:23:10 +0000

privleap (3:4.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 14 Nov 2025 06:56:19 +0000

privleap (3:4.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 01 Nov 2025 08:17:38 +0000

privleap (3:4.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 19 Sep 2025 16:16:02 +0000

privleap (3:4.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 02 Sep 2025 15:22:52 +0000

privleap (3:4.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 27 Aug 2025 19:28:34 +0000

privleap (3:3.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 26 Aug 2025 09:13:30 +0000

privleap (3:3.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 14 Jun 2025 12:00:35 +0000

privleap (3:3.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 11 Jun 2025 19:15:06 +0000

privleap (3:3.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 20 May 2025 11:39:46 +0000

privleap (3:3.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 07 May 2025 06:40:22 +0000

privleap (3:3.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 19 Apr 2025 17:33:12 +0000

privleap (3:3.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 09 Apr 2025 15:15:39 +0000

privleap (3:3.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 08 Apr 2025 14:08:04 +0000

privleap (3:3.1-1) unstable; urgency=medium

  * New upstream version (local package).


usr/lib/python3/dist-packages/privleap/leapctl.py
#!/usr/bin/python3 -su

# Copyright (C) 2025 - 2026 ENCRYPTED SUPPORT LLC <adrelanos@whonix.org>
# See the file COPYING for copying conditions.

# pylint: disable=broad-exception-caught,duplicate-code
# Rationale:
#   broad-exception-caught: except blocks are general error handlers.
#   duplicate-code: This is being triggered because of generic_error and
#     unexpected_error_msg, which are very similar in leapctl and leaprun but
#     can't be reasonably broken out of either due to the fact that they access
#     incompatible variants of the cleanup_and_exit function.

"""leapctl.py - privleapd client for controlling available comm sockets."""

import sys
import time
import pwd
from typing import NoReturn

from .privleap import (
    PrivleapCommon,
    PrivleapControlClientCreateMsg,
    PrivleapControlClientDestroyMsg,
    PrivleapControlClientReloadMsg,
    PrivleapControlServerControlErrorMsg,
    PrivleapControlServerDisallowedUserMsg,
    PrivleapControlServerExistsMsg,
    PrivleapControlServerExpectedDisallowedUserMsg,
    PrivleapControlServerNouserMsg,
    PrivleapControlServerOkMsg,
    PrivleapControlServerPersistentUserMsg,
    PrivleapMsg,
    PrivleapSession,
)


# pylint: disable=too-few-public-methods
# Rationale:
#   too-few-public-methods: This class just stores global variables, it needs no
#     public methods. Namespacing global variables in a class makes things
#     safer.
class LeapctlGlobal:
    """
    Global variables for leapctl.
    """

    control_session: PrivleapSession | None = None


def cleanup_and_exit(exit_code: int) -> NoReturn:
    """
    Ensures any active session is closed, then exits.
    """

    if LeapctlGlobal.control_session is not None:
        LeapctlGlobal.control_session.close_session()
    sys.exit(exit_code)


def print_usage() -> NoReturn:
    """
    Prints usage information.
    """

    print(
        """leapctl <--create|--destroy> <user>
leapctl --reload

     --create : Specifies that leapctl should request a communication socket to
                be created for the specified user account.
    --destroy : Specifies that leapctl should request a communication socket
                to be destroyed for the specified user account.
     --reload : Instructs privleapd to reload configuration without restarting.
         user : The username or UID of the user account to create or destroy a
                communication socket for."""
    )
    cleanup_and_exit(1)


def generic_error(error_msg: str, exit_int: int = 1) -> NoReturn:
    """
    Prints an error, then cleans up and exits.
    """

    print(f"ERROR: {error_msg}", file=sys.stderr)
    cleanup_and_exit(exit_int)


def unexpected_msg_error(control_msg: PrivleapMsg) -> NoReturn:
    """
    Alerts about an unexpected message type being received from the server.
    """

    print(
        "ERROR: privleapd returned an unexpected message as follows:",
        file=sys.stderr,
    )
    print(control_msg.serialize(), file=sys.stderr)
    cleanup_and_exit(1)


def wait_for_control_socket() -> None:
    """
    Waits for up to 5 seconds for the control socket to become available.
    """

    ## NOTE: We use polling to detect the socket. In nearly all cases, the
    ## socket will already exist, thus polling will only have to check for the
    ## socket's existence once, which is likely very fast. The only known case
    ## where the socket doesn't exist when leapctl is run is when the user
    ## runs something like `sudo systemctl restart privleapd; leapctl ...`, in
    ## which case minor delays from polling are acceptable. If in the future
    ## polling becomes a bottleneck, we may be able to use Varlink to listen
    ## for when the privleapd systemd unit comes up.

    for _ in range(50):
        if PrivleapCommon.control_path.is_socket():
            return
        time.sleep(0.1)


def start_control_session() -> None:
    """
    Starts a control session with the server.
    """

    try:
        LeapctlGlobal.control_session = PrivleapSession(is_control_session=True)
    except Exception:
        generic_error("Could not connect to privleapd!")


def handle_create_request(user_id: str) -> NoReturn:
    """
    Requests that privleapd create a socket for the specified user, and handles
    the response to the request.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        LeapctlGlobal.control_session.send_msg(
            PrivleapControlClientCreateMsg(user_id)
        )
    except Exception:
        generic_error("Could not request privleapd to create a comm socket!")

    try:
        control_msg: PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    if isinstance(control_msg, PrivleapControlServerOkMsg):
        print(f"Comm socket created for account {repr(user_id)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, PrivleapControlServerControlErrorMsg):
        generic_error(
            "privleapd encountered an error while creating a comm socket "
            f"for account {repr(user_id)}!"
        )
    elif isinstance(control_msg, PrivleapControlServerDisallowedUserMsg):
        generic_error(
            f"Account {repr(user_id)} is not permitted to have a comm socket!",
            2,
        )
    elif isinstance(
        control_msg, PrivleapControlServerExpectedDisallowedUserMsg
    ):
        print(
            f"Account {repr(user_id)} is not permitted to have a comm "
            "socket, as expected, ok."
        )
        cleanup_and_exit(0)
    elif isinstance(control_msg, PrivleapControlServerExistsMsg):
        print(f"Comm socket already exists for account {repr(user_id)}.")
        cleanup_and_exit(0)
    else:
        unexpected_msg_error(control_msg)


def handle_destroy_request(user_id: str) -> NoReturn:
    """
    Requests that privleapd destroy a socket for the specified user, and handles
    the response to the request.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        # noinspection PyUnboundLocalVariable
        LeapctlGlobal.control_session.send_msg(
            PrivleapControlClientDestroyMsg(user_id)
        )
    except Exception:
        generic_error("Could not request privleapd to destroy a comm socket!")

    try:
        control_msg: PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    if isinstance(control_msg, PrivleapControlServerOkMsg):
        print(f"Comm socket destroyed for account {repr(user_id)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, PrivleapControlServerControlErrorMsg):
        generic_error(
            "privleapd encountered an error while destroying a comm "
            f"socket for account {repr(user_id)}!"
        )
    elif isinstance(control_msg, PrivleapControlServerNouserMsg):
        print(f"Comm socket does not exist for account {repr(user_id)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, PrivleapControlServerPersistentUserMsg):
        print(f"Cannot destroy socket for persistent account {repr(user_id)}.")
        # It is not an error to try to destroy a socket for a persistent user,
        # since this may legitimately happen if someone logs in as a user that
        # happens to be persistent in privleap's config, and then logs out.
        cleanup_and_exit(0)
    else:
        unexpected_msg_error(control_msg)


def handle_reload_request() -> NoReturn:
    """
    Requests that privleapd reload its configuration files without restarting.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        LeapctlGlobal.control_session.send_msg(PrivleapControlClientReloadMsg())
    except Exception:
        generic_error("Could not request privleapd to reload configuration!")

    try:
        control_msg: PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    if isinstance(control_msg, PrivleapControlServerOkMsg):
        print("privleapd configuration reload successful.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, PrivleapControlServerControlErrorMsg):
        generic_error("privleapd failed to reload configuration!")
    else:
        unexpected_msg_error(control_msg)


def main() -> NoReturn:
    """
    Main function.
    """

    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print_usage()

    control_action: str = sys.argv[1]
    control_user_id: str | None = None
    if len(sys.argv) == 3:
        control_user_id = sys.argv[2]

    if control_action not in ("--create", "--destroy", "--reload"):
        print_usage()
    if control_action == "--reload" and control_user_id is not None:
        print_usage()

    if control_user_id is not None:
        control_user_struct: pwd.struct_passwd | None = (
            PrivleapCommon.normalize_user_id(control_user_id)
        )
        # Allow a username that doesn't exist to be passed when using --destroy,
        # so if a user is deleted before their comm socket is destroyed, the
        # socket can be destroyed anyway. Note that this will only work if the
        # deleted user's UID is passed, since there is no way to recover a
        # deleted user's UID to correlate it to a UID.
        if control_user_struct is None:
            if control_action != "--destroy":
                generic_error(
                    f"Specified account {repr(control_user_id)} does "
                    + "not exist."
                )
        else:
            control_user_id = str(control_user_struct.pw_name)

    wait_for_control_socket()
    start_control_session()

    if control_action == "--create":
        assert control_user_id is not None
        handle_create_request(control_user_id)
    elif control_action == "--destroy":
        assert control_user_id is not None
        handle_destroy_request(control_user_id)
    else:
        handle_reload_request()


if __name__ == "__main__":
    main()
